### PR TITLE
Inline protobuf definitions into collector repo

### PR DESCRIPTION
**Pull request by seanlinsley**
Historically the snapshot protobuf definitions were stored in a [separate repo](https://github.com/pganalyze/collector-snapshot/) so that the main application could include them as a submodule. But now that the the main application includes the collector itself as a submodule, the additional maintenance burden to have this as a separate repo no longer makes sense.

For reference, I followed [these steps](https://stackoverflow.com/a/1789374) to inline the protobuf definitions.

### CODE REVIEW

**Review by msakrejda — COMMENTED** [review]
+1 in general, but I wonder if instead we shouldn't
 1. create a branch in this repo
 2. drop the submodule in this repo
 3. move everything in the snapshot repo into a `protobuf` directory
 4. add the snapshot repo as a remote to this repo
 5. merge the snapshot repo main branch into this repo's branch

I haven't tried this yet, but I think that would let us preserve the full history of the snapshot repo. Thoughts?

In any case, after we merge this, we should remember to archive https://github.com/pganalyze/collector-snapshot to avoid confusion.

**Comment by seanlinsley** [comment]
Since the snapshot definitions are largely append-only and all of the actual implementation was in the collector repo, I'm not sure it's worth creating a confusing git history here just to preserve git blames.

Anecdotally, I can't think of a single time when I needed to know the provenance of a history field by looking at the protobuf definitions themselves; I'd always look at the collector repo or the main repo. And if we did merge the history of the two repos like you suggest, if I were to look at a git blame of the protobuf it wouldn't include the collector code changes since that'd still be a separate commit, so there wouldn't be a simple way to see all of the relevant changes.

**Comment by msakrejda** [comment]
I tried this locally (had to add `--allow-unrelated-histories`, but otherwise, it went smoothly), and the history looks pretty clean to me. All the protobuf history is basically just in the branch. You're right about `git blame`, but since the protobuf repo moves slowly, it shouldn't be hard to correlate by timestamp.

I'd still prefer this approach, but I'll let others weigh in.

**Comment by seanlinsley** [comment]
Let's walk through a possible scenario: you reach [this commit](https://github.com/pganalyze/collector-snapshot/commit/37171e773e957c09b4eed5a95ceb845b49e9f76e) from a git blame and want to find the related collector changes. Screenshot for reference:


I see two problems:

- The `#21` in the commit message would either no longer be a link, or it would incorrectly link to collector issue [#21](https://github.com/pganalyze/collector/issues/21).
- GitHub doesn't give you a way to find commits around the same timestamp as this commit; you can only navigate to its parent commit.

If the history was instead squashed, you'd reach this PR (#581) and then navigate to the [archived repo](https://github.com/pganalyze/collector-snapshot/), and perform the same git blame there. I think that leaves much less potential for confusion.

**Comment by msakrejda** [comment]
Thanks for illustrating the use case; that makes sense. I prefer to investigate history through `git log`, and I think that'd be easier with a combined history, but I concede the GH magic does not play well with that. I suspect external contributors would also prefer something GH-friendly, so maybe you're right. I'll leave it open for comment for a bit, but I think if no one chimes in, we can go with your approach.

**Comment by keiko713** [comment]
tl;dr I'm indecisive, and I'm okay with either way.

I was initially commenting here saying that we could go with Sean's approach, given the downside Sean illustrated + the history of will-be-archived collector-snapshot repo is likely going to be enough for me.
However, reading more about the stackoverflow post that Sean pasted in the PR summary (people in general wanting to preserve the history, including Maciek, and I do think it's a good practice too), also looking at the git blame UI of the current collector-snapshot, I wonder if the case Sean mentioned will be that common. IOW, if the downside is minimal, I want to support Maciek's way. That `#21` is squashed, and out of 30+ merged PR, I only see a few squashed ones... actually more like ~10 PRs squashed. The squashed ones will have that issue more indeed, as the commit message will have something like `(#21)`. Well, now I have not so small number in front of me, maybe the downside is not so minimal (also unsquashed ones still have the problem that it's not easy to find the PR from the commit).

Well, if we go with Sean's way, I may add one request. It'd be great if you can add something like "see https://github.com/pganalyze/collector-snapshot for previous history" to the commit message of this PR, so that when I do git blame or see the history of these files/directories in the future, I will have more easy pointer there.

**Review by msakrejda — APPROVED** [review]
Thanks for weighing in, @keiko713. I'm going to approve this because I realized that if we go with @seanlinsley's approach, we can always go back and merge histories later, and having this merged will simplify his other work.

+1 on mentioning the old snapshot repo in the commit message, though.
